lib/iaas/AzureClient.js

```diff
@@ -0,0 +1,192 @@
+    if (arguments.length < 2) {
+      options = container;
+      container = this.containerName;
+    }
+    const prefix = options ? options.prefix : null;
+    //TODO Need to update following line with value from 'options' e.g. options.marker
+    let continuationToken = null;
+    //TODO : maxResult is 5000 
+    //https://azure.github.io/azure-storage-node/BlobService.html#listBlobsSegmentedWithPrefix__anchor
+    //If want to fetch more use 'continuationToken' returned in result
+    return this.storage  
+      .listBlobsSegmentedWithPrefixAsync(container, prefix, continuationToken)  
+      .then(result => {
+        const listOfFiles = result.entries;
+        continuationToken = result.continuationToken;
+        let isTruncated = continuationToken ? true : false;         
+        const files = [];     
+        _.each(listOfFiles, file => files.push(_       
+          .chain(file)       
+          .pick('name', 'lastModified')       
+          .set('isTruncated', isTruncated)       
+          .value()          
+        ));      
+        return files;    
+      });
+  }
+
+  remove(container, file) {
+    if (arguments.length < 2) {
+      file = container;
+      container = this.containerName;
+    }
+    logger.debug(`Deleting file ${file} in container ${container} `);
+    return this.storage
+      .deleteBlobAsync(container, file)
+      .then(response => logger.info(`Response of blob deletion from cloud storage for ${file} :`, response))
+      .catch(BaseCloudClient.providerErrorTypes.Unauthorized, err => {
+        logger.error(err.message);
+        throw new Unauthorized('Authorization at the cloud storage provider failed');
+      })
+      .catch(BaseCloudClient.providerErrorTypes.Forbidden, err => {
+        logger.error(err.message);
+        throw new Forbidden('Authentication at the cloud storage provider failed');
+      })
+      .catchThrow(BaseCloudClient.providerErrorTypes.NotFound, new NotFound(`Object '${file}' not found`));
+  }
+
+  download(options) {
+    return utils.streamToPromise(this.storage.createReadStream(
+        options.container, options.remote))
+      .catchThrow(BaseCloudClient.providerErrorTypes.NotFound, new NotFound(`Object '${options.remote}' not found`));
+  }
+
+  upload(options, buffer) {
+    return new Promise((resolve, reject) => {
+      function cleanup() {
+        uploadedStream.removeListener('finish', onfinish);
+        uploadedStream.removeListener('error', onerror);
+      }
+
+      function onerror(err) {
+        cleanup();
+        reject(err);
+      }
+
+      function onfinish() {
+        cleanup();
+        resolve(JSON.parse(buffer));
+      }
+
+      let uploadedStream = this.storage
+        .createWriteStreamToBlockBlob(
+          options.container, options.remote);
+
+      uploadedStream.on('error', onerror);
+      uploadedStream.on('finish', onfinish);
+      uploadedStream.end(buffer);
+    });
+  }
+
+  uploadJson(container, file, data) {
```

Comment:
What is the difference between upload vs uploadJson? are both of them needed?

Reply:
Ideally  not required. uploadJson(): converts json to buffer and upload() uses only buffer. Is okay to have as someone may need to use upload() directly to feed a buffer when data not in json format.

Reply:
ok